File: src/vector_3D.cpp
/******************************************************************************
** Program : vector_3D.cpp
** Author  : Neil Massey
** Date    : 23/04/09
** Purpose : 3D vector class
******************************************************************************/

#include <assert.h>
#include <math.h>
#include "vector_3D.h"

vector_3D::vector_3D(void)
{
	x = y = z = 0.0;
}

/*****************************************************************************/

vector_3D::vector_3D(FP_TYPE px, FP_TYPE py, FP_TYPE pz)
{
	x = px;
	y = py;
	z = pz;
}

/*****************************************************************************/

vector_3D::vector_3D(const vector_3D& rhs)
{
	x = rhs.x;
	y = rhs.y;
	z = rhs.z;
}

/*****************************************************************************/

FP_TYPE& vector_3D::operator[](const int i)	
{
	assert(i >=0 && i < 3);
	switch(i)
	{
		case 0:
			return x;
		case 1:
			return y;
		case 2:
			return z;
	}
	return z;
}

/*****************************************************************************/

const FP_TYPE& vector_3D::operator[](const int i) const
{
	assert(i >=0 && i < 3);
	switch(i)
	{
		case 0:
			return x;
		case 1:
			return y;
		case 2:
			return z;
	}
	return z;
}

/*****************************************************************************/

vector_3D vector_3D::operator+(const vector_3D& rhs) const
{
	return vector_3D(x+rhs.x, y+rhs.y, z+rhs.z);
}

/*****************************************************************************/

vector_3D vector_3D::operator+(const FP_TYPE sc) const
{
	return vector_3D(sc+x, sc+y, sc+z);
}

/*****************************************************************************/

void vector_3D::operator+=(const vector_3D& rhs)
{
	x += rhs.x;
	y += rhs.y;
	z += rhs.z;
}

/*****************************************************************************/

void vector_3D::operator+=(const FP_TYPE sc)
{
	x += sc;
	y += sc;
	z += sc;
}

/*****************************************************************************/

vector_3D vector_3D::operator-(const vector_3D& rhs) const
{
	return vector_3D(x-rhs.x, y-rhs.y, z-rhs.z);
}

/*****************************************************************************/

vector_3D vector_3D::operator-(const FP_TYPE sc) const
{
	return vector_3D(x-sc, y-sc, z-sc);
}

/*****************************************************************************/

void vector_3D::operator-=(const vector_3D& rhs)
{
	x -= rhs.x;
	y -= rhs.y;
	z -= rhs.z;
}

/*****************************************************************************/

void vector_3D::operator-=(const FP_TYPE sc)
{
	x -= sc;
	y -= sc;
	z -= sc;
}

/*****************************************************************************/

vector_3D vector_3D::operator*(const FP_TYPE& sc) const
{
	return vector_3D(x*sc, y*sc, z*sc);
}

/*****************************************************************************/

void vector_3D::operator*=(const FP_TYPE& sc)
{
	x *= sc;
	y *= sc;
	z *= sc;
}

/*****************************************************************************/

vector_3D vector_3D::operator/(const FP_TYPE& sc) const
{
	return vector_3D(x/sc, y/sc, z/sc);
}

/*****************************************************************************/

void vector_3D::operator/=(const FP_TYPE& sc)
{
	x /= sc;
	y /= sc;
	z /= sc;
}

/*****************************************************************************/

bool vector_3D::operator==(const vector_3D& rhs) const
{
	return x == rhs.x && y == rhs.y && z == rhs.z;
}

/*****************************************************************************/

bool vector_3D::operator!=(const vector_3D& rhs) const
{
	return x != rhs.x && y != rhs.y && z != rhs.z;
}

/*****************************************************************************/

FP_TYPE vector_3D::mag(void) const
{
	FP_TYPE l = x*x + y*y + z*z;
	return sqrt(l);
}

/*****************************************************************************/

FP_TYPE vector_3D::sqmag(void) const
{
	FP_TYPE l = x*x + y*y + z*z;
	return l;
}

/*****************************************************************************/

FP_TYPE vector_3D::dp(const vector_3D& rhs) const
{
	return x*rhs.x + y*rhs.y + z*rhs.z;
}

/*****************************************************************************/

vector_3D vector_3D::xp(const vector_3D& rhs) const
{
	return vector_3D(y*rhs.z - z*rhs.y, 
					 z*rhs.x - x*rhs.z,
					 x*rhs.y - y*rhs.x);
}

/*****************************************************************************/

void vector_3D::zero(void)
{
	x = y = z = 0.0f;
}

/*****************************************************************************/

std::ostream& operator<<(std::ostream& os, const vector_3D& V)
{
	os << V.x << " " << V.y << " " << V.z;
	return os;
}

/*****************************************************************************/

std::istream& operator>>(std::istream& is, vector_3D& V)
{
	is >> V.x >> V.y >> V.z;
	return is;
}

File: python/vector_3D.py
#******************************************************************************
#** Program : vector_3D.py
#** Author  : Neil Massey
#** Date    : 26/07/13
#** Purpose : class to hold a 3D vector
#******************************************************************************

import numpy

class vector_3D:

	#**************************************************************************

	def __init__(self, xx=0, yy=0, zz=0):
		self.V = numpy.zeros(3, 'f');
		self.V[0] = xx
		self.V[1] = yy
		self.V[2] = zz

	#**************************************************************************

	def __getitem__(self, i):
		return self.V[i]
		
	#**************************************************************************
	
	def __setitem__(self, i, A):
		self.V[i] = A
	
	#**************************************************************************

	def __add__(self, rhs):
		new_vec = vector_3D()
		if isinstance(rhs, vector_3D):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] + rhs.V[i]
		elif isinstance(rhs, (int, float)):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] + rhs
		else:
			raise Exception("Unsupported right hand side type in operand")
		return new_vec

	#**************************************************************************

	def __sub__(self, rhs):
		new_vec = vector_3D()
		if isinstance(rhs, vector_3D):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] - rhs.V[i]
		elif isinstance(rhs, (int, float)):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] - rhs
		else:
			raise Exception("Unsupported right hand side type in operand")
		return new_vec

	#**************************************************************************

	def __mul__(self, rhs):
		new_vec = vector_3D()
		if isinstance(rhs, vector_3D):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] * rhs.V[i]
		elif isinstance(rhs, (int, float)):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] * rhs
		else:
			raise Exception("Unsupported right hand side type in operand")
		return new_vec

	#**************************************************************************

	def __div__(self, rhs):
		new_vec = vector_3D()
		if isinstance(rhs, vector_3D):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] / rhs.V[i]
		elif isinstance(rhs, (int, float)):
			for i in range(0,3):
				new_vec.V[i] = self.V[i] / rhs
		else:
			raise Exception("Unsupported right hand side type in operand")
		return new_vec

	#**************************************************************************

	def __eq__(self, rhs):
		if not (isinstance(rhs, vector_3D)):
			return False
		eq = True
		for i in range(0,3):
			eq &= (self.V[i] == rhs[i])
		return eq

	#**************************************************************************

	def __ne__(self, rhs):
		if not (isinstance(rhs, vector_3D)):
			return True
		eq = True
		for i in range(0,3):
			eq &= (self.V[i] != rhs[i])
		return eq

	#**************************************************************************

	def dp(self, rhs):
		# dot product
		return self.V[0]*rhs.V[0] + self.V[1]*rhs.V[1] + self.V[2]*rhs.V[2]

	#**************************************************************************

	def xp(self, rhs):
		return vector_3D(self.V[1]*rhs.V[2] - self.V[2]*rhs.V[1],
						 self.V[2]*rhs.V[0] - self.V[0]*rhs.V[2],
						 self.V[1]*rhs.V[0] - self.V[1]*rhs.V[0])

	#**************************************************************************

	def mag(self):
		return numpy.sqrt(self.V[0]*self.V[0] + self.V[1]*self.V[1] + self.V[2]*self.V[2])

	#**************************************************************************

	def sqmag(self):
		return (self.V[0]*self.V[0] + self.V[1]*self.V[1] + self.V[2]*self.V[2])

	#**************************************************************************
